test/functional/ui/incsubstitution_spec.lua

```diff
@@ -0,0 +1,1374 @@
+    for _, case in pairs(cases) do
+      refresh(case)
+      execute("cnoremap <expr> x cursor(1, 1)[-1].'x'")
+
+      feed(":%s/tw/tox/g<enter>")
+      expect(string.gsub(default_text, "tw", "tox"))
+    end
+  end)
+
+  it('work when a mapping disables incsub', function()
+    for _, case in pairs(cases) do
+      refresh(case)
+      execute("cnoremap <expr> x execute('set incsubstitute=')[-1]")
+
+      feed(":%s/tw/toxa/g<enter>")
+      expect(string.gsub(default_text, "tw", "toa"))
+    end
+  end)
+
+  it('work with a complex mapping', function()
+    for _, case in pairs(cases) do
+      refresh(case)
+      source([[cnoremap x <C-\>eextend(g:, {'fo': getcmdline()})
+      \.fo<CR><C-c>:new<CR>:bw!<CR>:<C-r>=remove(g:, 'fo')<CR>x]])
+
+      feed(":%s/tw/tox")
+      feed("/<enter>")
+      expect(string.gsub(default_text, "tw", "tox"))
+    end
+  end)
+
+end)
+
+describe('Incsubstitute: autocommands', function()
+  before_each(clear)
+
+  -- keys are events to be tested
+  -- values are arrays like 
+  --    { open = { 1 }, close = { 2, 3} } 
+  -- which would mean that during the test below the event fires for 
+  -- buffer 1 when opening an incsub window, and for buffers 2 and 3
+  -- when closing an incsub window
+  local eventsExpected = {
+    BufAdd = {},
+    BufDelete = { open = {2, 3} },
+    BufEnter = { open = {1, 1} },
+    BufFilePost = {},
+    BufFilePre = {},
+    BufHidden = { open = {2, 3} },
+    BufLeave = { open = {2, 3} },
+    BufNew = {},
+    BufNewFile = {},
+    BufRead = {},
+    BufReadCmd = {},
+    BufReadPre = {},
+    BufUnload = { open = {2, 3} },
+    BufWinEnter = {},
+    BufWinLeave = { open = {2, 3} },
+    BufWipeout = { open = {2, 3} },
+    BufWrite = {},
+    BufWriteCmd = {},
+    BufWritePost = {},
+    Syntax = {},
+    FileType = {},
+    WinEnter = { open = {1, 1} },
+    WinLeave = { open = {2, 3} },
+    CmdwinEnter = {},
+    CmdwinLeave = { open = {1, 1} },
+  }
+
+  local function bufferlist(t)
+    local s = ""
+    for _, buffer in pairs(t) do
+      s = s .. ", " .. tostring(buffer)
+    end
+    return s
+  end
+
+  -- fill the table with default values
+  for event, _ in pairs(eventsExpected) do
+    eventsExpected[event].open = eventsExpected[event].open or {}
+    eventsExpected[event].close = eventsExpected[event].close or {}
+  end
+
+
+  local function register_autocmd(event)
+    execute("let g:" .. event .. "_fired=[]")
+    execute("autocmd " .. event .. " * call add(g:" .. event .. "_fired, expand('<abuf>'))")
+  end
+
+  it('are run properly when splitting', function()
+    common_setup(nil, "split", default_text)
+
+    local eventsObserved = {}
+    for event, _ in pairs(eventsExpected) do
+      eventsObserved[event] = {}
+      register_autocmd(event)
+    end
+
+    feed(":%s/tw")
+
+    for event, _ in pairs(eventsExpected) do
+      eventsObserved[event].open = eval("g:" .. event .. "_fired")
+      execute("let g:" .. event .. "_fired=[]")
```

Comment:
IIRC the argument is "to simulate real user interaction" (which does not make sense in this case)